modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/ServerImpl.java

```diff
@@ -616,12 +625,80 @@ else if (log.isInfoEnabled()) {
         if (!res && node.clientRouterNodeId() == null && nodeAlive(nodeId)) {
             LT.warn(log, "Failed to ping node (status check will be initiated): " + nodeId);
 
-            msgWorker.addMessage(new TcpDiscoveryStatusCheckMessage(locNode, node.id()));
+            msgWorker.addMessage(createTcpDiscoveryStatusCheckMessage(locNode, locNode.id(), node.id()));
         }
 
         return res;
     }
 
+    /**
+     * Creates new instance of {@link TcpDiscoveryStatusCheckMessage} trying to choose most optimal constructor.
+     *
+     * @param creatorNode Creator node.
+     * @param creatorNodeId Creator node id, is used when <code>creatorNode</code> is null.
+     * @param failedNodeId Failed node id.
+     * @return new instance of {@link TcpDiscoveryStatusCheckMessage}.
+     */
+    private TcpDiscoveryStatusCheckMessage createTcpDiscoveryStatusCheckMessage(
+        @Nullable TcpDiscoveryNode creatorNode,
+        @Nullable UUID creatorNodeId,
+        UUID failedNodeId
+    ) {
+        assert creatorNode != null || creatorNodeId != null;
+
+        TcpDiscoveryStatusCheckMessage msg;
+
+        if (clusterSupportsTcpDiscoveryNodeSerializationOptimization) {
+            TcpDiscoveryNode crd = resolveCoordinator();
+
+            boolean sameMacs = creatorNode != null && crd != null && U.sameMacs(creatorNode, crd);
+
+            msg = new TcpDiscoveryStatusCheckMessage(
+                creatorNode == null ? creatorNodeId : creatorNode.id(),
+                spi.getNodeAddresses(creatorNode, sameMacs),
+                failedNodeId
+            );
+        }
+        else {
+            if (creatorNode != null)
+                msg = new TcpDiscoveryStatusCheckMessage(creatorNode, failedNodeId);
+            else
+                throw new IgniteSpiException("Cannot create status check message: creator node is unknown.");
```

Comment:
Also I have a doubt about this exception. In master we never have a situation when we fail to create a StatusCheckMessage.
This exception may mean that we changed semantics of our code. However I believe we cannot construct a scenario that triggers this branch, so please analyse thoroughly all situations when StatusCheck message could be created. My guess is that we can remove `@Nullable` annotation from `creatorNode` parameter and therefore remove branch throwing exception as well.

Reply:
We can receive null in creatorNode parameter when we construct TcpDiscoveryStatusCheckMessage from another message, as the point of optimization was not to serialize this creatorNode.
I replaced exception with returning null, but now I should check method result - if it is null, it means that creator node left topology, and we should do nothing, discarding the message.